ehsm_kms_service/ehsm_kms_server.js

```diff
@@ -0,0 +1,278 @@
+     {
+       code: int,
+       message: string,
+       result: {
+         cmk_base64 : string,
+       }
+    }
+   */
+  'NAPI_CreateKey': ['string',['int','int']],
+  /**
+    NAPI_Encrypt 
+    Description:
+     Encrypt an arbitrary set of bytes using the CMK.(only support symmetric types)
+
+    params
+     - cmk_base64: string
+     - plaintext: string
+     - aad: string
+    
+    return json
+     {
+       code: int,
+       message: string,
+       result: {
+         cipherText_base64 : string,
+       }
+    }
+   */
+  'NAPI_Encrypt': ['string',['string','string', 'string']],
+  
+  /**
+    NAPI_Decrypt 
+    Description:
+      Decrypts ciphertext using the CMK.(only support symmetric types)
+
+    params
+     - cmk_base64: string
+     - ciphertext: string
+     - aad: string
+    
+    return json
+     {
+       code: int,
+       message: string,
+       result: {
+         plaintext_base64 : string,
+       }
+    }
+   */
+  'NAPI_Decrypt': ['string',['string','string', 'string']],
+  /**
+    NAPI_GenerateDataKey 
+    Description:
+    Generates a random data key that is used to locally encrypt data.
+    the datakey will be wrapped by the specified CMK(only support asymmetric keyspec),
+    and it will return the plaintext and ciphertext of the data key.
+    You can use the plaintext of the data key to locally encrypt your data without using KMS
+    and store the encrypted data together with the ciphertext of the data key, then clear the
+    plaintext data from memory as soon as possible.
+    when you want to obtain the plaintext of datakey again, you can call the Decrypt with the
+    cmk to get the plaintext data.
+
+    params 
+     - cmk_base64: string
+     - aad: string
+     
+    return json
+     {
+       code: int,
+       message: string,
+       result: {
+         plaintext_base64 : string,
+         cipherText_base64 : string,
+       }
+    }
+   */
+  'NAPI_GenerateDataKey': ['string',['string','string']],
+
+});
+
+
+// base64 encode
+const base64_encode = (str) => new Buffer.from(str).toString('base64')
+// base64 decode
+const base64_decode =(base64_str) =>new Buffer.from(base64_str, 'base64').toString()
+
+/**
+ * ehsm napi result
+ * @param {function name} action 
+ * @param {} res 
+ * @param {NAPI_* function params} params 
+ * @returns 
+ */
+
+const napi_result = (action, res, params) => {
+  try {
+    // r : NAPI_(*) Return results
+    const r = JSON.parse(ehsm_napi[`NAPI_${action}`](...params));
+    if(action == apis.Decrypt && r.code == 200){
+      r.result.plaintext_base64 = base64_decode(r.result.plaintext_base64);
+      res.send(r)
+    } else {
+      res.send(r);
+    }
+  } catch (e) {
+    res.send(result(400, e))
+  }
+  return;
+}
+
+/**
+ * router
+ */
+app.post('/ehsm', function (req, res) {
+  const PAYLOAD = req.body.payload;
+  
+  // ACTION: request function name
+  const ACTION = req.query.Action;
+  const keyspec = ehsm_keyspec_t[PAYLOAD.keyspec];
```

Comment:
why not put this param (const keyspec = ehsm_keyspec_t[PAYLOAD.keyspec];) under 
if(ACTION === apis.CreateKey) {
...
}
?